controllers/reconcilers/kserve_authconfig_reconciler.go

```diff
@@ -0,0 +1,160 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconcilers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	kservev1beta1 "github.com/kserve/kserve/pkg/apis/serving/v1beta1"
+	authorinov1beta2 "github.com/kuadrant/authorino/api/v1beta2"
+	"github.com/opendatahub-io/odh-model-controller/controllers/comparators"
+	"github.com/opendatahub-io/odh-model-controller/controllers/processors"
+	"github.com/opendatahub-io/odh-model-controller/controllers/resources"
+	"github.com/pkg/errors"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type KserveAuthConfigReconciler struct {
+	client         client.Client
+	scheme         *runtime.Scheme
+	deltaProcessor processors.DeltaProcessor
+	detector       resources.AuthTypeDetector
+	store          resources.AuthConfigStore
+	templateLoader resources.AuthConfigTemplateLoader
+	hostExtractor  resources.InferenceServiceHostExtractor
+}
+
+func NewKserveAuthConfigReconciler(client client.Client, scheme *runtime.Scheme) *KserveAuthConfigReconciler {
+	return &KserveAuthConfigReconciler{
+		client:         client,
+		scheme:         scheme,
+		deltaProcessor: processors.NewDeltaProcessor(),
+		detector:       resources.NewKServeAuthTypeDetector(client),
+		store:          resources.NewClientAuthConfigStore(client),
+		templateLoader: resources.NewConfigMapTempalteLoader(client, resources.NewStaticTemplateLoader()),
+		hostExtractor:  resources.NewKServeInferenceServiceHostExtractor(),
+	}
+}
+
+func (r *KserveAuthConfigReconciler) Reconcile(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+
+	if isvc.Status.URL == nil {
+		log.V(1).Info("Inference Service not ready yet, waiting for URL")
+		return nil
+	}
+
+	log.V(1).Info("create desired state")
+	desiredState, err := r.createDesiredResource(ctx, isvc)
+	if err != nil {
+		return err
+	}
+
+	log.V(1).Info("get existing state")
+	existingState, err := r.getExistingResource(ctx, isvc)
+	if err != nil && !k8serror.IsNotFound(err) {
+		return err
+	}
+
+	// Process Delta
+	log.V(1).Info("process delta")
+	if err = r.processDelta(ctx, log, desiredState, existingState); err != nil {
+		return err
+	}
+	return nil
+
+}
+
+func (r *KserveAuthConfigReconciler) createDesiredResource(ctx context.Context, isvc *kservev1beta1.InferenceService) (*authorinov1beta2.AuthConfig, error) {
+	typeName := types.NamespacedName{
+		Name:      isvc.GetName(),
+		Namespace: isvc.GetNamespace(),
+	}
+
+	authType, err := r.detector.Detect(ctx, isvc)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not detect AuthType for InferenceService %s", typeName)
```

Comment:
@aslakknutsen @bartoszmajsak The unmaintained libray... uhm :confused:

Reply:
In this repo it is going to be the first usage. Indeed, Filippe pointed that this library now would become a direct dependency [here](https://github.com/opendatahub-io/odh-model-controller/pull/130#discussion_r1420924906).

I think, I'd prefer to not introduce it as a dependency, unless we have a good reason to (directly) depend on that 3rd party library.

Reply:
https://github.com/opendatahub-io/odh-model-controller/pull/130/commits/5c1fdf8a99e58492a17508406d6e79e7875a22bc